bin/node/build.rs

```diff
@@ -0,0 +1,95 @@
+use std::path::PathBuf;
+
+use anyhow::{Context, Result};
+
+/// Configures environment variables for build metadata intended for extended version information.
+fn main() -> Result<()> {
+    if let Some(sha) = published_git_sha() {
+        // git data is not available if in a published state, so we set them manually.
+        println!("cargo::rustc-env=VERGEN_GIT_SHA={sha}");
+        println!("cargo::rustc-env=VERGEN_GIT_BRANCH=NA (published)");
+        println!("cargo::rustc-env=VERGEN_GIT_DIRTY=");
+
+        vergen_gitcl::Emitter::new()
+    } else {
+        // In a non-published state so we can expect git instructions to work.
+        let mut emitter = vergen_gitcl::Emitter::new();
+        emitter
+            .add_instructions(&git_instructions()?)
+            .context("Adding git instructions")?;
+
+        emitter
+    }
+    .add_instructions(&cargo_instructions()?)
+    .context("Adding cargo instructions")?
+    .add_instructions(&rustc_instructions()?)
+    .context("Adding rustc instructions")?
+    .emit()
+}
+
+/// Normal git info is lost on `cargo publish`, which instead adds a file containing the SHA1 hash.
+///
+/// This function returns the short SHA value. If present, this indicates this we're in a published
+/// state.
+fn published_git_sha() -> Option<String> {
+    let cargo_vcs_info = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join(".cargo_vcs_info.json");
+    if cargo_vcs_info.exists() {
+        // The file is small so reading to string is acceptable.
+        let contents = std::fs::read_to_string(cargo_vcs_info).expect("Reading vcs info");
+
+        // File format:
+        // {
+        //   "git": {
+        //     "sha1": "9d48046e9654d93a86212e77d6c92f14c95de44b"
+        //   },
+        //   "path_in_vcs": "bin/node"
+        // }
+        let offset =
+            contents.find(r#""sha1""#).expect("Searching for sha1 property") + r#""sha1""#.len();
+
+        let sha1 = contents[offset + 1..]
+            .chars()
+            // Find and skip opening quote.
+            .skip_while(|&c| c != '"')
+            .skip(1)
+            // Take until closing quote.
+            .take_while(|&c| c != '"')
+            // Short SHA format is 7 digits.
+            .take(7)
+            .collect();
+
+        Some(sha1)
+    } else {
+        None
+    }
```

Comment:
Would it make sense to make this a stand alone function that lives in the utils crate? This way, at least the file parsing logic would be shared between different binaries.

Reply:
I moved everything instead.